src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1493,19 +1493,24 @@
 
 ;; Wrap the fact-type so that Clojure equality and hashcode semantics are used
 ;; even though this is placed in a Java map.
-(deftype AlphaRootsWrapper [roots roots-hash]
+(deftype AlphaRootsWrapper [fact-type fact-type-hash roots]
   Object
   (equals [this other]
-    ;; Since we memoize the function that retrieves these on the fact type, and since the alpha roots for
-    ;; any given type primary type T (excluding descendants and ancestors) are a disjoint
-    ;; set with respect to the alpha roots for any other primary type T2, we know that we will
-    ;; only have one AlphaRootsWrapper object for any given set of alpha roots per get-alphas-fn instance.
-    ;; Therefore we can just use reference equality in the interest of maximizing performance.
-    (identical? this other))
+    (let [other ^AlphaRootsWrapper other]
+      (cond
+        
+        (identical? fact-type (.fact-type other)) 
+        true
+        
+        (not (== fact-type-hash (.fact-type-hash other)))
```

Comment:
Summary: I've investigated and I'm OK with this change now.

When you add the int type hint to the fact-type-hash field of the AlphaRootsWrapper, AOT the clara.rules.compiler namespace by adding

````
:aot [clara.rules.compiler]
````

to the project.clj and decompile the AlphaRootsWrapper to Java syntax with JD-GUI, the constructor of the AlphaRootsWrapper looks like this:

````
public AlphaRootsWrapper(Object paramObject1, int paramInt, Object paramObject2)
  {
    this.fact_type = paramObject1;this.fact_type_hash = paramInt;this.roots = paramObject2;
  }
````

I then tried decompiling interop calls to the AlphaRootsWrapper constructor and found that the second argument to the constructor was wrapped in clojure.lang.RT/intCast [1].  The call looks like 

````
return new AlphaRootsWrapper(null, RT.intCast(x), null);
````

The Clojure compiler inserts this code at [2].  If \*unboxed-math\* is false, the default, a call to RT.intCast is inserted.  If \*unboxed-math\* is true then a call to RT.uncheckedIntCast [3] is inserted.  If the parameter type is not a primitive, then the JVM opcode for a checked cast is inserted. To validate my understanding of this I cross-referenced the 192 base 10 opcode the compiler inserts against [4]. In either case it looks like passing either boxed types or primitives should be safe.  This is why building an AlphaRootsWrapper like this works, with \*unchecked-math\* set to both true and false:

````
(AlphaRootsWrapper. nil (Integer/valueOf 1) nil)
````

Note that the first and third arguments to the AlphaRootsWrapper wouldn't actually be nil, I just didn't care about them for the purposes of this investigation.

1. https://github.com/clojure/clojure/blob/clojure-1.8.0/src/jvm/clojure/lang/RT.java#L1169
2. https://github.com/clojure/clojure/blob/clojure-1.8.0/src/jvm/clojure/lang/Compiler.java#L893
3. https://github.com/clojure/clojure/blob/clojure-1.8.0/src/jvm/clojure/lang/RT.java#L1429
4. https://en.wikipedia.org/wiki/Java_bytecode_instruction_listings